File: basic/median_of_two_sorted_arrays.java
public class Median_of_two_sorted_arrays {
    public static double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int len1 = nums1.length;
        int len2 = nums2.length;
        int size = len1 + len2;
        if(size % 2 == 1)
            return findKth(nums1,  0, len1, nums2, 0, len2, size/2 + 1);
        else
            return (findKth(nums1, 0, len1, nums2, 0, len2, size/2)
                  + findKth(nums1, 0, len1, nums2, 0, len2, size/2 + 1)) / 2;
    }

    private static double findKth(int[] nums1, int start1, int len1,
                            int[] nums2, int start2, int len2, int k)
    {
        if(len1 - start1 > len2 - start2) // 传进来的时候统一让短的数组为nums1
            return findKth(nums2, start2, len2, nums1, start1, len1, k);
        if(len1 - start1 == 0) // 表示nums1已经全部加入前k个了，第k个为nums2[k-1];
            return nums2[k - 1];
        if(k == 1) // k==1表示已经找到第k-1小的数，下一个数为两个数组start开始的最小值
            return Math.min(nums1[start1], nums2[start2]);
        int p1 = start1 + Math.min(len1 - start1, k/2); // p1和p2记录当前需要比较的那个位
        int p2 = start2 + k - p1 + start1;
        System.out.printf("p1,p2 %d %d\n", p1,p2);
        if(nums1[p1 - 1] < nums2[p2 - 1])
            return findKth(nums1, p1, len1, nums2, start2, len2, k - p1 + start1);
        else if(nums1[p1 - 1] > nums2[p2 - 1])
            return findKth(nums1, start1, len1, nums2, p2, len2, k - p2 + start2);
        else
            return nums1[p1 - 1];
    }

    private static double findKth_v2(int[] nums1, int start1, int len1,
                            int[] nums2, int start2, int len2, int k)
    {
        if(len1 - start1 > len2 - start2) // 传进来的时候统一让短的数组为nums1
            return findKth(nums2, start2, len2, nums1, start1, len1, k);
        if(len1 - start1 == 0) // 表示nums1已经全部加入前k个了，第k个为nums2[k-1];
            return nums2[k - 1];
        if(k == 1) // k==1表示已经找到第k-1小的数，下一个数为两个数组start开始的最小值
            return Math.min(nums1[start1], nums2[start2]);
        int p1 = start1 + Math.min(len1 - start1, k/2); // p1和p2记录当前需要比较的那个位
        int p2 = start2 + k - p1 + start1;
        System.out.printf("p1,p2 %d %d\n", p1,p2);
        if(nums1[p1 - 1] < nums2[p2 - 1])
            return findKth(nums1, p1, len1, nums2, start2, len2, k - p1 + start1);
        else if(nums1[p1 - 1] > nums2[p2 - 1])
            return findKth(nums1, start1, len1, nums2, p2, len2, k - p2 + start2);
        else
            return nums1[p1 - 1];
    }

    public static void main(String[] args) {
        int[] A = {1,2,3,10,11,12};
        int[] B = {4,5,6,7,8,9};
        double median = Median_of_two_sorted_arrays.findMedianSortedArrays(A,B);
        System.out.printf("median:%f",median);
    }
}

File: Basic/median_of_two_sorted_arrays.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int len1 = nums1.size();
        int len2 = nums2.size();
        if ((len1 + len2) % 2 == 1)
            return findKth((len1+len2)/2+1, nums1, 0, nums2, 0);
        else
            return  (findKth((len1+len2)/2, nums1, 0, nums2, 0)
                   + findKth((len1+len2)/2 + 1, nums1, 0, nums2, 0)) / 2;
    }

    double findKth(int k, vector<int>&A, int offset1, vector<int>B, int offset2)
    {
        // cout << "K:" << k << " A:" << offset1 << " " << A.size() << " B:" << offset2 << " " << B.size() << endl;
        if (A.size()-offset1 > B.size()-offset2)
        {
            // cout << A.size()-offset1 << ", " <<  B.size()-offset2 << " Swap A and B" << endl;
            return findKth(k, B, offset2, A, offset1);
        }

        if (offset1 == A.size())
            return B[k-1];
        if (k == 1)
            return min(A[offset1], B[offset2]);

        int p1 = min(A.size()-offset1, k/2);
        int p2 = k - p1;

        if (A[offset1+p1-1] == B[offset2+p2-1])
            return A[offset1+p1-1];
        else if (A[offset1+p1-1] < B[offset2+p2-1]) // discard left part of A
            return findKth(k-p1, A, offset1+p1, B, offset2);
        else // discard left part of B
            return findKth(k-p2, A, offset1, B, offset2+p2);
    }
    int min(int a, int b){return a<b? a:b;}
};

int main(int argc, char const *argv[])
{
    Solution sln;
    vector<int> n1,n2;
    n1.push_back(1),n1.push_back(2),n1.push_back(3),n1.push_back(4);
    n2.push_back(5),n2.push_back(6),n2.push_back(7),n2.push_back(8);
    cout << sln.findMedianSortedArrays(n1,n2);
    return 0;
}